include/swift/Reflection/ReflectionContext.h

```diff
@@ -111,6 +114,98 @@ class ReflectionContext
+    auto SectionHdrAddress = ImageStart.getAddressData() + Hdr->e_shoff;
+    auto SectionHdrNumEntries = Hdr->e_shnum;
+    auto SectionEntrySize = Hdr->e_shentsize;
+
+    // Collect all the section headers, we need them to look up the
+    // reflection sections (by name) and the string table.
+    std::vector<const Elf64_Shdr *> SecHdrVec;
+    for (unsigned I = 0; I < SectionHdrNumEntries; ++I) {
+      auto SecBuf = this->getReader().readBytes(
+	RemoteAddress(SectionHdrAddress +  (I * SectionEntrySize)),
+	SectionEntrySize);
+      auto SecHdr = reinterpret_cast<const Elf64_Shdr *>(SecBuf.get());
+      SecHdrVec.push_back(SecHdr);
+    }
+ 
+    // This provides quick access to the section header string table index.
+    // We also here handle the unlikely even where the section index overflows
+    // and it's just a pointer to secondary storage (SHN_XINDEX).
+    uint32_t SecIdx = Hdr->e_shstrndx;
+    if (SecIdx == SHN_XINDEX) {
+      assert(!SecHdrVec.empty() && "malformed ELF object");
+      SecIdx = SecHdrVec[0]->sh_link;
+    }
+
+    assert(SecIdx < SecHdrVec.size() && "malformed ELF object");
+
+    const Elf64_Shdr *SecHdrStrTab = SecHdrVec[SecIdx];
+    Elf64_Off StrTabOffset = SecHdrStrTab->sh_offset;
+    Elf64_Xword StrTabSize = SecHdrStrTab->sh_size;
+
+    auto StrTabStart = RemoteAddress(ImageStart.getAddressData() + StrTabOffset);
+    auto StrTabBuf = this->getReader().readBytes(StrTabStart, StrTabSize);
+    auto StrTab = reinterpret_cast<const char *>(StrTabBuf.get());
+
+    auto findELFSectionByName = [&](std::string Name) -> std::pair<const char *,
+								   const char *>{
+      // Now for all the sections, find their name.
+      for (const Elf64_Shdr *Hdr : SecHdrVec) {
+        uint32_t Offset = Hdr->sh_name;
+        auto SecName = std::string(StrTab + Offset);
+        if (SecName != Name)
+          continue;
+        auto SecStart = RemoteAddress(ImageStart.getAddressData() + Hdr->sh_offset);
+        auto SecSize = Hdr->sh_size;
+        auto SecBuf = this->getReader().readBytes(SecStart, SecSize);
+	auto SecContents = reinterpret_cast<const char *>(SecBuf.get());
+        return {SecContents, SecContents + SecSize};
+      }
+      return {nullptr, nullptr};
+    };
+
+    // FIXME: there's no swift5_builtin and swift5_capture section on Linux.
+    auto FieldMdSec = findELFSectionByName("swift5_fieldmd");
+    if (FieldMdSec.first == nullptr || FieldMdSec.second == nullptr)
+      return false;
+    auto AssocTySec = findELFSectionByName("swift5_assocty");
+    if (AssocTySec.first == nullptr || AssocTySec.second == nullptr)
+      return false;
+    auto TypeRefMdSec = findELFSectionByName("swift5_typeref");
+    if (TypeRefMdSec.first == nullptr || TypeRefMdSec.second == nullptr)
+      return false;
+    auto ReflStrMdSec = findELFSectionByName("swift5_reflstr");
+    if (ReflStrMdSec.first == nullptr || ReflStrMdSec.second == nullptr)
+      return false;
```

Comment:
For Mach-O, we succeed if *any* of the sections are found, as at least some of them are optional. I think ELF should probably do the same. That would also allow you to have this code read swift5_builtin and swift5_capture, and then it will start reading those whenever the compiler starts emitting them.

Reply:
I think the semantics that Mach-O gives are fine (for now), but we might want to revisit them in the future.
If, e.g. for some reason the compiler doesn't emit the `typeref` section or the `reflstr` section then we're in big trouble because the debugger kind-of relies on them to be present. I imagine in that case you might at least emit a warning and degrade gracefully in lldb, but I also understand this is a discussion for another day.

[... or similarly, somebody runs the executable through strip or objcopy which drops these sections on the floor]

Reply:
I guess it depends on whether you want best effort or all-or-nothing. Might depend on the API’s client. I think some of these sections really are optional, at least I’ve seen freshly built binaries that don’t have them all.

Reply:
If the compiler has nothing to emit for any of the sections, it probably ends up being omitted entirely from the binary. For that reason, I think that it's best to consider them all as optional, or treat the associated section as effectively empty if it doesn't exist in the binary in question. We should treat the lack of a section when the debugger expects it as a format error further down at the point we find out we need it, IMO.